File: 2025/day_1.py
def get_dial_current_position(turn: str, start: int = 0, _cycle: int = 100) -> tuple[int, int]:
    """
    This performs modular arithmetic and returns a tuple with the first element being the 
    current dial after performing the turns the second element is the number of times
    the dial crossed 0
    """
    if _cycle == 0:
        raise ValueError("The Cycle can not be Zero")
    
    # "R10"[:1] 
    # LX WHERE X is a integer
    
    turn_steps = int(turn[1:])
    direction = -1 if turn.startswith("L") else 1    
    new_position = (start + (direction * turn_steps)) % _cycle 
    
    crossings = 0
    for i in range(turn_steps):
        current_value = (start + (direction * i)) % _cycle
        if current_value == 0 and (i != 0 or i > (turn_steps - 1)):
            crossings += 1
        
    return (new_position, crossings)
    
    
def get_total_zero_hit(seq: list[str], start: int = 0, _cycle: int = 100) -> int:
    total_zero_hit = 0
    for turn in seq:
        current_dial, zero_count = get_dial_current_position(turn, start, _cycle)
        start = current_dial
        
        if current_dial == 0:
            total_zero_hit += 1
            
        total_zero_hit += zero_count
        
    return total_zero_hit



if __name__ == "__main__":
    seq = open("2025/data/day_1_input.txt").readlines()
    seq = [turn.strip() for turn in seq]
    result = get_total_zero_hit(seq=seq, start=50)
    print(f"Total Zero Hit: {result}")
    

File: 2025/day_1.rs
use std::fs;

fn get_dial_current_position(turn: String, start: u32, _cycle: u32) -> (u32, u32) {
    if _cycle == 0 {
        panic!("The Cycle Can not be Zero");
    }
    
    let turn_steps = turn[1..].trim().parse::<u32>().unwrap();
    let direction = if turn.starts_with("L") { -1 } else { 1 };
    let new_position = (start as i32 + (direction * turn_steps as i32)) % _cycle as i32;
    
    let mut crossings = 0;
    for i in 0..turn_steps {
        let current_value = (start as i32 + (direction * i as i32)) % _cycle as i32;
        if current_value == 0 && (i != 0 || i > (turn_steps - 1)) {
            crossings += 1;
        }
    }
    
    (new_position as u32, crossings as u32)
}

fn get_total_zero_hit(seq: Vec<String>, start: u32, _cycle: u32) -> u32 {
    let mut start: u32 = start;
    let mut total_zero_hit: u32 = 0;
    for turn in seq.into_iter() {
        let (current_dial, zero_count) = get_dial_current_position(turn, start, _cycle);
        start = current_dial;
        
        if current_dial == 0 {
            total_zero_hit += 1;
        }
        
        total_zero_hit += zero_count;
    }
    
    return total_zero_hit;
}

fn main() {
    let binding = fs::read_to_string("2025/data/day_1_input.txt").expect("Could not read from file");
    let lines = binding.lines();
    let lines: Vec<String> = lines.map(|l| l.to_owned()).collect();
        
    let result = get_total_zero_hit(lines, 50, 100);
    println!("Total Zero Hit: {:?}", result);
}